File: Week 7/id_520/52.n皇后-ii.java
/*
 * @lc app=leetcode.cn id=52 lang=java
 *
 * [52] N皇后 II
 */

// @lc code=start
class Solution {
    int size;
    int count;

    public void solve(int row, int pie, int na) {
        //terminator
        if(row == size) {
            count++;
            return ;
        }
        //当前得到的所有空位
        int pos = size & (~(row | pie | na));
        while(pos != 0) {
            // p 为当前pos最低位的1
            int p = pos & (-pos);
            //更新pos，把pos最低位的1置零，代表在这上面放了皇后
            pos &= pos - 1;
            // drill down
            solve(row | p, (pie | p) << 1, (na | p) >> 1 );
        }

    }
    public int totalNQueens(int n) {
        count = 0;
        // size表示n位前面都置零的一个算子，但是这里也用于了终止循环，很巧妙
        size = ( 1 << n) - 1;
        solve(0, 0, 0);
        return count;
    }
}
// @lc code=end

File: Week 7/id_445/52.n皇后-ii.cpp
/*
 * @lc app=leetcode.cn id=52 lang=cpp
 *
 * [52] N皇后 II
 */

// @lc code=start
class Solution
{
public:
    int totalNQueens(int n)
    {
        dfs(n,0,0,0,0);
        return res;
    }

    void dfs(int n, int row, int col, int ld, int rd)
    {
        if (row >= n)
        {
            res++;
            return;
        }

        // 将所有能放置 Q 的位置由 0 变成 1，以便进行后续的位遍历
        int bits = ~(col | ld | rd) & ((1 << n) - 1);
        while (bits > 0)
        {
            int pick = bits & -bits; // 注: x & -x
            dfs(n, row + 1, col | pick, (ld | pick) << 1, (rd | pick) >> 1);
            bits &= bits - 1; // 注: x & (x - 1)
        }
    }

private:
    int res = 0;
};
// @lc code=end
